### DISCUSSION TITLE
Question about ParamSpec Changes in 1.1.359

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#7730] [Q&A] [answered]
Post 1.1.359 being released, I'm now seeing errors with this usage of `ParamSpec` in the context of narrowing for `**kwargs` when it's annotated with `P.kwargs`. Here's a contrived example:

```python
from typing import Callable, Concatenate, Generic, ParamSpec

P = ParamSpec("P")

class Container(Generic[P]):
    def __init__(
        self,
        func: Callable[Concatenate[int, P], str],
    ) -> None:
        self.func = func

    def __call__(
        self,
        val: int,
        *args: P.args,
        **kwargs: P.kwargs,
    ) -> str:
        if "data" in kwargs:
            raise ValueError("data is not allowed in kwargs")

        return self.func(val, *args, **kwargs)
```

[Playground link.](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDSFgoDGRM1KrdUA4u9SEib0ACkRBEIAZQTUmAKDnCoAXiijxUmUwAUAImG6AlArlNyAZ3MMUMIqn7beKfoIDawgLqGAXHKj%2BoABNqYCgAfTDUJBgI7T8AhPNqEmBaeIT-YABXZm9CUnIqalcCRhY2DjZXVBgRD3pzGBB69KhDKABaAD4oADlGal8MgKSUgDps5hUoSflW4NCIljJY1sTk1LWAgDdSPJq04f9qAA8mojCxNDyKMDASQ6OAKivzPOEx18fhp6eAawA7q93mNAV9Wu1ulBGiAhkckKFdIFWERdJgUFAwSA0G8tsNxEgklAAGqkLLUACiIHAID0yNsmCsKDA%2BAKYAB1EC6MxQOx5iMCiOUBA1BgWRAGNGwAmOR0uweUBefPovyxOMMQA)

You might be asking, why do this seemingly redundant check? Mostly for the sake of non type-checked legacy code. There's sort of this transformation of the call signature through a decorator, and if you weren't paying attention to the new type determined by the decorator, you might think you have to pass a specific key-word argument. If you're using Pyright of course, you can immediately tell you what's going on, but not all engineers have this configured, so they could miss it.

By explicitly checking if "data" is in `kwargs`, it gets narrows to `dict[str, object]`, causing an error when trying to return the encapsulated callable: `Arguments for ParamSpec "P@Container" are missing`

Was it intentional that any narrowing of args/kwargs in the context of param spec means you can no longer pass them into the annotated function? If so, I can understand why, since really you shouldn't be doing any additional checks on it like this, if your codebase was fully annotated. But just want to make sure before upgrading.

**Comment by erictraut** [comment, marked as answer]
This wasn't intentional. It was the side effect of a different change. I'll create a bug report from this discussion.

**Reply by erictraut** [reply]
https://github.com/microsoft/pyright/issues/7731

**Reply by max-muoto** [reply]
Thanks Eric, really appreciate it!
